runtime/compiler/optimizer/BoolArrayStoreTransformer.cpp

```diff
@@ -0,0 +1,434 @@
+
+   /*
+    * Do a DFS of the CFG as the best effort to figure out types in one traverse
+    */
+   TR::BlockChecklist visitedBlocks(comp());
+   blockStack.push(comp()->getStartBlock());
+   while (!blockStack.empty() && !_bstoreiUnknownArrayTypeNodes->empty())
+      {
+      TR::Block *block = blockStack.top();
+      blockStack.pop();
+      visitedBlocks.add(block);
+      TypeInfo *blockEndTypeInfo = processBlock(block, blockStartTypeInfos[block->getNumber()]);
+      TR_SuccessorIterator bi(block);
+      for (TR::CFGEdge *edge = bi.getFirst(); edge != NULL; edge = bi.getNext())
+         {
+         TR::Block *nextBlock = toBlock(edge->getTo());
+         if (!visitedBlocks.contains(nextBlock))
+            {
+            blockStack.push(nextBlock);
+            //propagate auto type info to successor
+            if (blockEndTypeInfo)
+               blockStartTypeInfos[nextBlock->getNumber()] = new (currentRegion) TypeInfo(*blockEndTypeInfo);
+            }
+         }
+      }
+   }
+
+/*
+ * Answer includes multi dimension boolean array
+ */
+bool TR_BoolArrayStoreTransformer::isAnyDimensionBoolArrayNode(TR::Node *node)
+   {
+   return getArrayDimension(node, true /*boolType*/) >= 1;
+   }
+
+/*
+ * Answer includes multi dimension byte array
+ */
+bool TR_BoolArrayStoreTransformer::isAnyDimensionByteArrayNode(TR::Node *node)
+   {
+   return getArrayDimension(node, false /*boolType*/) >= 1;
+   }
+
+/*
+ * Answer includes one dimension boolean array only
+ */
+bool TR_BoolArrayStoreTransformer::isBoolArrayNode(TR::Node *node)
+   {
+   return getArrayDimension(node, true /*boolType*/) == 1;
+   }
+
+/*
+ * Answer includes one dimension byte array only
+ */
+bool TR_BoolArrayStoreTransformer::isByteArrayNode(TR::Node *node)
+   {
+   return getArrayDimension(node, false /*boolType*/) == 1;
+   }
+
+/*
+ * \brief
+ *     Get dimension of the array of given type
+ *
+ * \parm node
+ *    The node to look at
+ *
+ * \parm boolType
+ *    True if asking for boolean array and false if asking for byte array
+ *
+ * \return
+ *    The dimension of the array of given type. Return -1 if the node is not the
+ *    given type.
+ */
+int TR_BoolArrayStoreTransformer::getArrayDimension(TR::Node *node, bool boolType)
+   {
+   int nodeArrayDimension = -1;
+   if (node->getOpCodeValue() == TR::newarray)
+      {
+      int32_t expectedTypeValue = boolType ? 4: 8;
+      TR::Node *arrayTypeNode = node->getSecondChild();
+      TR_ASSERT(arrayTypeNode->getOpCode().isLoadConst(), "expect the second child of TR::newarray to be constant " POINTER_PRINTF_FORMAT, arrayTypeNode);
+      if (arrayTypeNode->getInt() == expectedTypeValue)
+         nodeArrayDimension = 1;
+      }
+   else
+      {
+      char expectedTypeChar = boolType? 'Z' : 'B';
+      int length;
+      const char * signature = node->getTypeSignature(length);
+      if (signature && length >= 2 && signature[length-1] == expectedTypeChar && signature[length-2] == '[')
```

Comment:
added checking `length >= 2` and check if the second last char must be `[` to make sure the type is indeed array or multi array.